narwhals/functions.py

```diff
@@ -88,15 +100,58 @@ def from_dict(data: dict[str, Any], *, native_namespace: Any) -> DataFrame[Any]:
         └─────┴─────┘
     """
     if native_namespace is get_polars():
-        native_frame = native_namespace.from_dict(data)
+        if schema:
+            from narwhals._polars.utils import (
+                reverse_translate_dtype as polars_reverse_translate_dtype,
+            )
+
+            schema = {
+                name: polars_reverse_translate_dtype(dtype)
+                for name, dtype in schema.items()
+            }
+
+        native_frame = native_namespace.from_dict(data, schema=schema)
     elif (
         native_namespace is get_cudf()
         or native_namespace is get_modin()
         or native_namespace is get_pandas()
     ):
         native_frame = native_namespace.DataFrame.from_dict(data)
+
+        if schema:
+            from narwhals._pandas_like.utils import Implementation
+            from narwhals._pandas_like.utils import (
+                reverse_translate_dtype as pandas_like_reverse_translate_dtype,
+            )
+
+            implementation = (
+                Implementation.PANDAS
+                if native_namespace is get_pandas()
+                else Implementation.MODIN
+                if native_namespace is get_modin()
+                else Implementation.CUDF
+            )
+            schema = {
+                name: pandas_like_reverse_translate_dtype(
+                    schema[name], native_type, implementation
+                )
+                for name, native_type in native_frame.dtypes.items()
+            }
+            native_frame = native_frame.astype(schema)
```

Comment:
could we do the `astype` on each column before creating `native_frame`? it might save pandas from consolidating columns into 2D blocks which we might then need to split up if `schema` assigns them different dtypes

Reply:
How would this work in practice? Meaning, I would expect that `data` is a dictionary which values can be something like python lists and I would not know how to cast those to (dataframe) native types before having the dataframe. What am I missing?

Reply:
could you call `pd.Series(..., copy=False, dtype=dtype)` on each one?

`copy=False` should be fine as we don't have any in-place operations in the Narwhals API

Reply:
actually, what you've done is fine